### PR TITLE
Implement celery with redis and flower monitoring

### LINKED ISSUE #90: Implement celery to handle UI while waiting for long api endpoints

### PR DESCRIPTION
**Pull request by avpeery**
### Why:
* The app has a couple jobs that take a bit of time to complete. Implementing celery with redis as broker and using flower as an http api to monitor the task statuses allows for a better user experience, such as the progress bar to demonstrate loading.

### What:
* Updated the readme to reflect changes to how to run the app locally with redis, celery and flower. Please follow these steps to test out the work flow and install missing dependencies. Celery is currently only being used on CBA tool processing, but will be easy to implement with load forecast now that the setup is there. Progress bar created in react with already installed material-ui to help with user experience.

_p.s.  Sorry for opening the PR Friday evening, but just wanted to get this done before the weekend. Please take until next week to look it over. Thanks!_

demo

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -379,4 +379,6 @@ $RECYCLE.BIN/
 webserver/script/costbenefitanalysis/cases/test_files
 webserver/script/costbenefitanalysis/preprocessing_loadprofiles/inputs/weekdays/*
 webserver/script/costbenefitanalysis/preprocessing_loadprofiles/inputs/weekends/*
-webserver/script/costbenefitanalysis/preprocessing_loadprofiles/outputs/*
\ No newline at end of file
+webserver/script/costbenefitanalysis/preprocessing_loadprofiles/outputs/*
+webserver/app/redis-6.0.8
+webserver/app/redis-6.0.8.tar.gz
```
Comment (jongoncalves) [inline .gitignore]: this git ignore entry is a bit weird, doesn't hurt anything, but having the binary shouldn't be in the project dir in the first place.

Comment (avpeery) [reply .gitignore]: Thanks for the insight! I’ll re-evaluate how I’m installing Redis and revise this.

File: README.md [hunk 2]
```diff
@@ -35,13 +35,27 @@ brew services start postgresql
...
 # install the backend dependencies
 cd webserver
 pip install -r requirements.txt
+
 # other parts of the project requires other dependencies....
 # instead of going through them and figuring out that you need those dependencies
 # when your code breaks, just install them now:
-pip install celery pandas cvxpy sklearn
+pip install celery==4.4.7
+pip install pandas cvxpy sklearn matplotlib s3fs flower redis xlrd
```
Comment (jongoncalves) [inline README.md:47]: Flower is a good approach!
We don't need to worry too much about this right now, but using flower complicates deployment a bit. Worth having a conversation about this later.

File: frontend/src/Components/AlgInputs/AlgInputs.js [hunk 3]
```diff
@@ -42,6 +42,7 @@ export default function AlgInputs (props) {
                 <AlgInputCustomized
                     category = { props.category }
                     visualizeResults={ (result) => visualizeResults(result) }
+                    loadingResults={ (bool) => props.loadingResults(bool)}
```
Comment (jongoncalves) [inline frontend/src/Components/AlgInputs/AlgInputs.js]: naming this argument `bool` is confusing. Something more descriptive like `isLoading` would be better

File: frontend/src/Components/AlgInputs/AlgInputsCBA.js [hunk 4]
```diff
@@ -114,16 +115,41 @@ class AlgInputsCBA extends Component {
...
-                url: postUrl,
-                data: {load_profile: this.state.profileName, county: countyMatch},
+                data: {load_profile: this.state.profileName, county: countyMatch}
+            
             });
+
+            // celery-flower monitoring to check task status on cba tool
+            let status = cba_res.data.status;
+            const task_id = cba_res.data.task_id;
+            while (status !== "SUCCESS") {
```
Comment (jongoncalves) [inline frontend/src/Components/AlgInputs/AlgInputsCBA.js]: There's an assumption here that the task will always succeed. 
I think if the task fails, we should stop the loop. I also think we should implement a simple exponential backoff algorithm, something like: https://jsfiddle.net/pajtai/pLka0ow9

File: frontend/src/Components/AlgorithmPage/AlgorithmPage.js [hunk 5]
```diff
@@ -1,27 +1,31 @@
...
 class AlgorithmPage extends Component {
     constructor(props) {
         super(props);
         this.state = {
             results: [],
+            loading: false
         };
     }
 
+    loadingResults(bool) {
```
Comment (jongoncalves) [inline frontend/src/Components/AlgorithmPage/AlgorithmPage.js]: same argument naming comment as the other.

File: webserver/app/celery.py [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+app = Celery('tasks', broker='redis://localhost:6379/0', backend='redis://localhost')
+
+app.config_from_object('django.conf:settings')
+
+# Load task modules from all registered Django app configs.
+app.autodiscover_tasks(settings.INSTALLED_APPS)
+
+@app.task(bind=True)
+def debug_task(self):
+    print(f'Request: {self.request!r}')
```
Comment (jongoncalves) [inline webserver/app/celery.py]: We can remove this debug task right?

File: frontend/src/Components/AlgInputs/AlgInputs.js (hunk 3 above)
Comment (avpeery) [reply frontend/src/Components/AlgInputs/AlgInputs.js]: fixed

File: frontend/src/Components/AlgInputs/AlgInputsCBA.js (hunk 4 above)
Comment (avpeery) [reply frontend/src/Components/AlgInputs/AlgInputsCBA.js]: fixed

File: frontend/src/Components/AlgorithmPage/AlgorithmPage.js (hunk 5 above)
Comment (avpeery) [reply frontend/src/Components/AlgorithmPage/AlgorithmPage.js]: fixed

File: webserver/app/celery.py (hunk 6 above)
Comment (avpeery) [reply webserver/app/celery.py]: fixed

**Comment by avpeery** [comment]
@derins @jongoncalves  ready for review. Will do different PR to add celery to load forecast page.

**Comment by avpeery** [comment]
Here is a picture of the current alert if the CBA tool celery task fails: 
Screen Shot 2020-10-23 at 2 44 05 PM

File: frontend/src/Components/Helpers/helpers.js [hunk 7]
```diff
@@ -64,3 +66,27 @@ export function preprocessData(allData) {
     }
     return resultFlattened;
 }
+
+export async function checkFlowerTaskStatus (task_id) {
+    const task_res = await axios({
+        url: `http://localhost:5555/api/task/result/${task_id}`,
+        method: "get"
+    });
+    return task_res.data.state
```
Comment (jongoncalves) [inline frontend/src/Components/Helpers/helpers.js]: missing `;`

File: frontend/src/Components/Helpers/helpers.js [hunk 8]
```diff
@@ -64,3 +66,27 @@ export function preprocessData(allData) {
...
+    if (status === "SUCCESS") {
+        callback();
+    } else {
+        if(status === "FAILURE" || max === 0){
+            clearTimeout(timeout);
+            return "FAILURE";
+        } else if (status === "PENDING") {
+            timeout = setTimeout(function() {
+                return exponentialBackoff(checkStatus, task_id, timeout, --max, delay * 2, callback);
+            }, delay)
```
Comment (jongoncalves) [inline frontend/src/Components/Helpers/helpers.js]: missing `;`

**Review by jongoncalves — APPROVED** [review]

**Review by derins — APPROVED** [review]

**Pull request merged**